### PR TITLE
[Sofa.Component.ODESolver] Gather tests + create Sofa.Component.ODESolver.Testing

### PR DESCRIPTION
**Pull request by fredroy**
All in the title.

~~Sofa.Testing.Tools aims to regroup all the test templates/skeletons, etc, so the *_test could use it.
It should also a "header-only" (aka INTERFACE syntax in cmake) but for now it is treated like a (empty) shared lib.
By itself it does not manage any dependencies, it is the role of the *_test itself to handle its own dependencies.~~

Sofa.Component.ODESolver.Testing gathers all tools/test template for odesolver tests.
Next step should removing SofaCreator and use SimpleAPI instead (and thus removing all binaries dependencies)

But the finalization of this should be reviewed/discussed.

Mandatory ref #1527
______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: SofaKernel/modules/Sofa.Testing/tools/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+set(SOFATESTINGTOOLS_SRC "src/sofa/testing/tools")
+
+# only header files
+set(HEADER_FILES
+    ${SOFATESTINGTOOLS_SRC}/MassSpringSystemCreation.h
+    ${SOFATESTINGTOOLS_SRC}/SunPlanetSystemCreation.h
+)
+
+set(SOURCE_FILES
+    ${SOFATESTINGTOOLS_SRC}/empty.cpp # because of not interface type
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Testing/tools/CMakeLists.txt]: What's the reason this project is not an INTERFACE?

Comment (fredroy) [reply SofaKernel/modules/Sofa.Testing/tools/CMakeLists.txt]: SOFA macros does not support the INTERFACE keyword

File: SofaKernel/modules/SofaExplicitOdeSolver/SofaExplicitOdeSolver_test/CMakeLists.txt [hunk 2]
```diff
@@ -3,11 +3,9 @@ cmake_minimum_required(VERSION 3.12)
 project(SofaExplicitOdeSolver_test)
 
 set(HEADER_FILES
-    MassSpringSystemCreation.h
 )
 
-set(SOURCE_FILES
-    EulerExplicitSolverDynamic_test.cpp)
+set(SOURCE_FILES)
```
Comment (guparan) [inline SofaKernel/modules/SofaExplicitOdeSolver/SofaExplicitOdeSolver_test/CMakeLists.txt:8]: File to be removed?

Comment (guparan) [inline SofaKernel/modules/SofaExplicitOdeSolver/SofaExplicitOdeSolver_test/CMakeLists.txt]: Remove SofaExplicitOdeSolver_test completely?

File: SofaKernel/modules/SofaImplicitOdeSolver/SofaImplicitOdeSolver_test/CMakeLists.txt [hunk 3]
```diff
@@ -2,14 +2,8 @@ cmake_minimum_required(VERSION 3.12)
 
 project(SofaImplicitOdeSolver_test)
 
-set(SOURCE_FILES
-    loadPlugins.cpp
-    EulerImplicitSolverStatic_test.cpp
-    EulerImplicitSolverDynamic_test.cpp
-    SpringSolverDynamic_test.cpp
-    StaticSolver_test.cpp)
+set(SOURCE_FILES)
```
Comment (guparan) [inline SofaKernel/modules/SofaImplicitOdeSolver/SofaImplicitOdeSolver_test/CMakeLists.txt]: Remove SofaImplicitOdeSolver_test completely?

**Comment by damienmarchal** [comment]
Hi @fredroy 

Several questions: 
- shouldn't we drop the dependency to SceneCreator (as discussed (https://github.com/sofa-framework/sofa/issues/417) ? 
- why not defining the test in Sofa.component.ODESolver.tests in place of Sofa.component.ODESolver_test ? To me this looks more consistant and coherent with our naming scheme as well as having the tests in a child directory of ODESolver.   
- why centralizing the testing.tools which breaks modularity instead of letting them at their right logical location ? 
To me having something like
```console
SofaDeformable.testing.tools/MassSpringSystemCreation   
``` 
Is much better than
```console
Sofa.testing.tools/MassSpringSystemCreation   
``` 
Because the latter creates a dependency from package SofaDeformable  to package Sofa.testing.toos. 
While the First has only a dependency from SofaDeformable to SofaDeformable.testing

**Comment by fredroy** [comment]
> Hi @fredroy
> 
> Several questions:
> 
>     * shouldn't we drop the dependency to SceneCreator (as discussed ([SceneCreator dependencies #417](https://github.com/sofa-framework/sofa/issues/417)) ?
> 
See https://github.com/sofa-framework/sofa/pull/2692#issuecomment-1040462403

>     * why not defining the test in Sofa.component.ODESolver.tests in place of Sofa.component.ODESolver_test ? To me this looks more consistant and coherent with our naming scheme as well as having the tests in a child directory of ODESolver.

Hum I kept the naming from before, and this is an executable, not a library so maybe something to differenciate. But I dont really care the naming, and we can change it if needed.

> 
>     * why centralizing the testing.tools which breaks modularity instead of letting them at their right logical location ?
>       To me having something like
> 
> 
> ```
> SofaDeformable.testing.tools/MassSpringSystemCreation   
> ```
> 
> Is much better than
> 
> ```
> Sofa.testing.tools/MassSpringSystemCreation   
> ```

For those two (`MassSpringSystemCreation` and `SunSystemCreation`), it is not really creating something ODESolver-specific so there is no "logical" place by itself.

**Comment by fredroy** [comment]
[ci-build][force-full-build][with-all-tests]

**Comment by damienmarchal** [comment]
> >  why not defining the test in Sofa.component.ODESolver.tests in place of Sofa.component.ODESolver_test ? To me this looks more consistant and coherent with our naming scheme as well as having the tests in a child directory of ODESolver.

> Hum I kept the naming from before, and this is an executable, not a library so maybe something to differenciate. But I dont really care the naming, and we can change it if needed.

I wouldn't say it is just name because we have given a specific meaning to the '.'. Sofa.component.ODESolver.tests is a subpackage (and a subdir) of Sofa.component.ODESolver while Sofa.component.ODESolver_tests is a side package. 
Shouldn't we prefer to have the tests shipped with the code they are supposed to validate ? 

> > * why centralizing the testing.tools which breaks modularity instead of letting them at their right logical location ?
> >   To me having something like
> > ```
> > SofaDeformable.testing.tools/MassSpringSystemCreation   
> > ```
> > Is much better than
> > ```
> > Sofa.testing.tools/MassSpringSystemCreation   
> > ```

> For those two (`MassSpringSystemCreation` and `SunSystemCreation`), it is not really creating something ODESolver-specific so there is no "logical" place by itself.

The given example is nothing related to ODESolver but SofaDeformable. As MassSpringSystemCreation is creating  StiffSpringForceField (which is in SofaDeformable), then the logical place seems to straightforwardly point to SofaDeformable and not Sofa.testing.whatever.

**Pull request merged**